TIMESTAMP_TRUNCATED_TO_MONTH & TIMESTAMP_TRUNCATED_TO_YEAR

### Feature description

Add more truncated timestamp group by options:
TIMESTAMP_TRUNCATED_TO_WEEK
TIMESTAMP_TRUNCATED_TO_MONTH
TIMESTAMP_TRUNCATED_TO_YEAR

### Motivating example

Clearly missing from the list:
TIMESTAMP_TRUNCATED_TO_HOUR
TIMESTAMP_TRUNCATED_TO_DAY

This is required for my reporting project. Until this is in, I'll have to manually build queries for these.

This is an incredible tool to use for reporting services on an isolated reporting replica... it only makes sense to complete this list so that we can group our data in all the meaningful ways, using a terse built-in method that already exists.

### Breaking changes

I don't see any reason this would cause breaking changes, simply extending current features.

### Supporting development

I [tick all that apply]:

<!-- To tick a checkbox, change it from `[ ]` to `[x]` -->

- [ ] am interested in building this feature myself
- [x] am interested in collaborating on building this feature
- [x] am willing to help testing this feature before it's released
- [ ] am willing to write a test-driven test suite for this feature (before it exists)
- [ ] am a [Graphile sponsor](https://www.graphile.org/sponsor/) ❤️
- [ ] have an active [support or consultancy contract](https://www.graphile.org/support/) with Graphile

> If the plugin isn't found or causes an error, is something output? Because when I intentionally use an invalid path, everything runs without warnings.

So it turns out you actually _can_ pass the function directly! https://github.com/graphile/postgraphile/blob/e0eeb1a1f768f7c757d36a6fa91045add92f9169/src/postgraphile/cli.ts#L596 Sorry for the misinformation here, it's a long time since I've visited this code!

> If the plugin isn't found or causes an error, is something output? Because when I intentionally use an invalid path, everything runs without warnings.

Should be! 
- https://github.com/graphile/postgraphile/blob/e0eeb1a1f768f7c757d36a6fa91045add92f9169/src/postgraphile/cli.ts#L610-L614
- https://github.com/graphile/postgraphile/blob/e0eeb1a1f768f7c757d36a6fa91045add92f9169/src/postgraphile/cli.ts#L619-L621 
- https://github.com/graphile/postgraphile/blob/e0eeb1a1f768f7c757d36a6fa91045add92f9169/src/postgraphile/cli.ts#L628-L630

Kinda indicates your config file is not actually being used? I wonder if you're editing the wrong file or running postgraphile in the wrong folder? Or maybe you're editing a `.postgraphilerc.ts` file and then not compiling it? Here's what happens when I run PostGraphile with a postgraphilerc with an appendPlugins that doesn't exist:

```
benjie@Benjie-Box:~/Dev/DELETEME/jeremybradbury$ cat .postgraphilerc.js 
module.exports = {
  options: {
    appendPlugins: [
      'does-not-exist'
    ]
  }
};
benjie@Benjie-Box:~/Dev/DELETEME/jeremybradbury$ postgraphile -c test
Failed to load plugin 'does-not-exist'
/home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js:287
            throw e;
            ^

Error: Cannot find module 'does-not-exist'
Require stack:
- /home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js
- /home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:927:15)
    at Function.Module._load (node:internal/modules/cjs/loader:772:27)
    at Module.require (node:internal/modules/cjs/loader:999:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at /home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js:282:20
    at Array.map (<anonymous>)
    at loadPlugins (/home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js:274:18)
    at Object.<anonymous> (/home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js:342:23)
    at Module._compile (node:internal/modules/cjs/loader:1095:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1124:10) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/build/postgraphile/cli.js',
    '/home/benjie/.nvm/versions/node/v16.4.0/lib/node_modules/postgraphile/cli.js'
  ]
}
```

> and tracking down append plugin errors (including fully missing files) currently seems fuzzy at best.

I almost never use `.postgraphilerc.js` and it's being removed/replaced in the next version with something better, it's been deprecated for a long time. That said the code seems to indicate that an error should be thrown if requiring fails, or if the plugin is of the wrong type, and that certainly seems to be the case in the quick test I did above.

I generally recommend that people using a `.postgraphilerc.js` move to using PostGraphile in library mode, it's just a few more lines of code and makes debugging a lot more straightforward because it's much more explicit what you are/are not running.

> I tried this variation too, based on the source document:

That one should definitely be throwing an error because there _should_ be a conflict on the `pgAggregateGroupBySpecs` name.

Also if you just add `throw new Error("TESTING");` to the root of the file you should see that error aborting startup and exiting node.

> There may be an opinionated reason not to accept it

There's a couple:

- I don't want to add too many enum values to people's schemas, especially if they don't need that feature
- I don't want to add "ambiguous" group by values, e.g. "week", where someone might want it grouped by week starting on Monday and someone else by week starting on Sunday. People can opt into that, but it's their choice. Same goes for month because they might want something slightly different (e.g. group by "every 4 weeks" so that they get samples of similar sizes), and adding hour, day and year seems weird without the others.

Hopefully we can make it sufficiently easy to add your own group by (maybe even with your code as an example in the README) that we don't need to add it to core. And of course you could always create a `graphile-aggregate-more-groups` or similar plugin that adds more groupings so people can `--append-plugins @graphile/pg-aggregates,graphile-aggregate-more-groups`.

--- a/README.md
+++ b/README.md
@@ -341,21 +341,91 @@ details/examples.
 
 ## Defining your own grouping derivatives
 
-You may add your own derivatives by adding a group by spec; see
-[src/AggregateSpecsPlugin.ts](src/AggregateSpecsPlugin.ts) for examples and more
-information. Derivative specs are also fairly straightforward:
+You may add your own derivatives by adding a group by spec to
+`build.pgAggregateGroupBySpecs` via a plugin. Derivative specs are fairly
+straightforward, for example here's the spec for "truncated-to-hour":
 
 ```ts
 const TIMESTAMP_OID = "1114";
 const TIMESTAMPTZ_OID = "1184";
 
 const truncatedToHourSpec = {
+  // A unique identifier for this spec, will be used to generate its name:
   id: "truncated-to-hour",
+
+  // A filter to determine which column/function return types this derivative
+  // is valid against:
   isSuitableType: (pgType) =>
     pgType.id === TIMESTAMP_OID || pgType.id === TIMESTAMPTZ_OID,
+
+  // The actual derivative - given the SQL fragment `sqlFrag` which represents
+  // the column/function call, return a new SQL fragment that represents the
+  // derived value, in this case a truncated timestamp:
   sqlWrap: (sqlFrag) => sql.fragment`date_trunc('hour', ${sqlFrag})`,
 };
 ```
+
+Building that up with a few more different intervals into a full PostGraphile
+plugin, you might write something like:
+
+```ts
+// Constants from PostgreSQL
+const TIMESTAMP_OID = "1114";
+const TIMESTAMPTZ_OID = "1184";
+
+// Determine if a given type is a timestamp/timestamptz
+const isTimestamp = (pgType) =>
+  pgType.id === TIMESTAMP_OID || pgType.id === TIMESTAMPTZ_OID;
+
+// Build a spec that truncates to the given interval
+const tsTruncateSpec = (sql, interval) => ({
+  // `id` has to be unique, derive it from the `interval`:
+  id: `truncated-to-${interval}`,
+
+  // Only apply to timestamp fields:
+  isSuitableType: isTimestamp,
+
+  // Given the column value represented by the SQL fragment `sqlFrag`, wrap it
+  // with a `date_trunc()` call, passing the relevant interval.
+  sqlWrap: (sqlFrag) =>
+    sql.fragment`date_trunc(${sql.literal(interval)}, ${sqlFrag})`,
+});
+
+// This is the PostGraphile plugin; see:
+// https://www.graphile.org/postgraphile/extending/
+const DateTruncAggregateGroupSpecsPlugin = (builder) => {
+  builder.hook("build", (build) => {
+    const { pgSql: sql } = build;
+
+    build.pgAggregateGroupBySpecs = [
+      // Copy all existing specs, except the ones we're replacing
+      ...build.pgAggregateGroupBySpecs.filter(
+        (spec) => !["truncated-to-day", "truncated-to-hour"].includes(spec.id)
+      ),
+
+      // Add our timestamp specs
+      tsTruncateSpec(sql, "year"),
+      tsTruncateSpec(sql, "month"),
+      tsTruncateSpec(sql, "week"),
+      tsTruncateSpec(sql, "day"),
+      tsTruncateSpec(sql, "hour"),
+      // Other values: microseconds, milliseconds, second, minute, quarter,
+      // decade, century, millennium.
+      // See https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC
+    ];
+
+    return build;
+  });
+};
+
+module.exports = DateTruncAggregateGroupSpecsPlugin;
+```
+
+Finally pass this plugin into PostGraphile via `--append-plugins` or
+`appendPlugins: [...]` - see https://www.graphile.org/postgraphile/extending/
+
+See [src/AggregateSpecsPlugin.ts](src/AggregateSpecsPlugin.ts) for examples and
+more information.
 
 ## Thanks
 

--- a/__tests__/date_trunc_aggregate_group_specs_plugin.js
+++ b/__tests__/date_trunc_aggregate_group_specs_plugin.js
@@ -1,0 +1,50 @@
+// Constants from PostgreSQL
+const TIMESTAMP_OID = "1114";
+const TIMESTAMPTZ_OID = "1184";
+
+// Determine if a given type is a timestamp/timestamptz
+const isTimestamp = (pgType) =>
+  pgType.id === TIMESTAMP_OID || pgType.id === TIMESTAMPTZ_OID;
+
+// Build a spec that truncates to the given interval
+const tsTruncateSpec = (sql, interval) => ({
+  // `id` has to be unique, derive it from the `interval`:
+  id: `truncated-to-${interval}`,
+
+  // Only apply to timestamp fields:
+  isSuitableType: isTimestamp,
+
+  // Given the column value represented by the SQL fragment `sqlFrag`, wrap it
+  // with a `date_trunc()` call, passing the relevant interval.
+  sqlWrap: (sqlFrag) =>
+    sql.fragment`date_trunc(${sql.literal(interval)}, ${sqlFrag})`,
+});
+
+// This is a PostGraphile plugin; see
+// https://www.graphile.org/postgraphile/extending/
+const DateTruncAggregateGroupSpecsPlugin = (builder) => {
+  builder.hook("build", (build) => {
+    const { pgSql: sql } = build;
+
+    build.pgAggregateGroupBySpecs = [
+      // Copy all existing specs, except the ones we're replacing
+      ...build.pgAggregateGroupBySpecs.filter(
+        (spec) => !["truncated-to-day", "truncated-to-hour"].includes(spec.id)
+      ),
+
+      // Add our timestamp specs
+      tsTruncateSpec(sql, "year"),
+      tsTruncateSpec(sql, "month"),
+      tsTruncateSpec(sql, "week"),
+      tsTruncateSpec(sql, "day"),
+      tsTruncateSpec(sql, "hour"),
+      // Other values: microseconds, milliseconds, second, minute, quarter,
+      // decade, century, millennium.
+      // See https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC
+    ];
+
+    return build;
+  });
+};
+
+module.exports = DateTruncAggregateGroupSpecsPlugin;

--- a/postgraphile.sh
+++ b/postgraphile.sh
@@ -2,7 +2,7 @@ node \
   --inspect \
   ../build/postgraphile/cli.js \
   --append-plugins \
-    `pwd`/node_modules/postgraphile-plugin-connection-filter,`pwd`/dist/index.js \
+    `pwd`/node_modules/postgraphile-plugin-connection-filter,`pwd`/dist/index.js,`pwd`/__tests__/date_trunc_aggregate_group_specs_plugin.js \
   -c graphile_aggregates \
   -s test \
   --enhance-graphiql \

